common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h

```diff
@@ -54,10 +56,67 @@ class mon_wlan_hal_dwpal : public base_wlan_hal_dwpal, public mon_wlan_hal {
 
     // Private data-members:
 private:
+    bool dwpal_get_scan_params_fg(sScanCfgParams &params)
+    {
+        if (sizeof(params) != dwpal_nl_cmd_get(m_radio_info.iface_name,
+                                               LTQ_NL80211_VENDOR_SUBCMD_GET_SCAN_PARAMS,
+                                               m_nl_buffer, NL_MAX_REPLY_BUFFSIZE)) {
+            LOG(ERROR) << "LTQ_NL80211_VENDOR_SUBCMD_GET_SCAN_PARAMS failed! returned size does "
+                          "not match sScanCfgParams!";
+            return false;
+        }
+
+        if (memcpy_s(&params, sizeof(params), &m_nl_buffer[NL_ATTR_HDR], sizeof(params)) != 0) {
+            LOG(ERROR) << __func__ << " memcpy_s failed!";
+            return false;
+        }
+        return true;
+    }
+
+    bool dwpal_get_scan_params_bg(sScanCfgParamsBG &params)
+    {
+        if (sizeof(params) != dwpal_nl_cmd_get(m_radio_info.iface_name,
+                                               LTQ_NL80211_VENDOR_SUBCMD_GET_SCAN_PARAMS_BG,
+                                               m_nl_buffer, NL_MAX_REPLY_BUFFSIZE)) {
+            LOG(ERROR) << "LTQ_NL80211_VENDOR_SUBCMD_GET_SCAN_PARAMS_BG failed! returned size does "
+                          "not match sScanCfgParamsBG!";
+            return false;
+        }
+
+        if (memcpy_s(&params, sizeof(params), &m_nl_buffer[NL_ATTR_HDR], sizeof(params)) != 0) {
```

Comment:
This triggers a build failure: https://gitlab.com/prpl-foundation/prplMesh/-/jobs/401484296

> `mmemcpy_s` was not declared in this scope

If you're going to use safelibc functions in the header file, you have to include the header (which still requires the ifdeffery to support both rdkb and ugw/openwrt).

However, I'd suggest moving it to the .cpp file.

And while you're at it, use std::copy maybe?

However, there's definitely something wrong here. You check that the return value of dwpal_nl_cmd_get (which is the length of the returned buffer) is equal to sizeof(params), and then you copy out sizeof(params) with an offset of 4 (NL_ATTR_HEADER). So you're guaranteeing an out of bounds access. So either you need to add NL_ATTR_HDR to the check above, or you have to copy without offset.

If the offset is needed, it is going to be needed for all calls of dwpal_nl_cmd_get. So you'd better skip the header inside the dwpal_nl_cmd_get function then. But then, the copy is again redundant, you can just pass params and sizeof(params) as arguments to dwpal_nl_cmd_get.

In conclusion, I don't think this copy will be needed.

BTW I have the feeling I've already made this comment in the original #619.

Reply:
Good thing we have gitlab CI!

Reply:
Currently building the image and running it is done in a single step. Building the image is really verbose (and we want to keep this verbosity to debug failing builds) and would flood the CI logs completely.
Once #621 is done, we can probably output everything when running the image only (i.e. when building prplmesh itself).

Reply:
> This triggers a build failure: https://gitlab.com/prpl-foundation/prplMesh/-/jobs/401484296
> 
> > `mmemcpy_s` was not declared in this scope
> 
> If you're going to use safelibc functions in the header file, you have to include the header (which still requires the ifdeffery to support both rdkb and ugw/openwrt).
> 
> However, I'd suggest moving it to the .cpp file.
> 
> And while you're at it, use std::copy maybe?
> 
> However, there's definitely something wrong here. You check that the return value of dwpal_nl_cmd_get (which is the length of the returned buffer) is equal to sizeof(params), and then you copy out sizeof(params) with an offset of 4 (NL_ATTR_HEADER). So you're guaranteeing an out of bounds access. So either you need to add NL_ATTR_HDR to the check above, or you have to copy without offset.
> 
> If the offset is needed, it is going to be needed for all calls of dwpal_nl_cmd_get. So you'd better skip the header inside the dwpal_nl_cmd_get function then. But then, the copy is again redundant, you can just pass params and sizeof(params) as arguments to dwpal_nl_cmd_get.
> 
> In conclusion, I don't think this copy will be needed.
> 
> BTW I have the feeling I've already made this comment in the original #619.



> This triggers a build failure: https://gitlab.com/prpl-foundation/prplMesh/-/jobs/401484296
> 
> > `mmemcpy_s` was not declared in this scope
> 
> If you're going to use safelibc functions in the header file, you have to include the header (which still requires the ifdeffery to support both rdkb and ugw/openwrt).
> 
> However, I'd suggest moving it to the .cpp file.
> 
> And while you're at it, use std::copy maybe?
> 
> However, there's definitely something wrong here. You check that the return value of dwpal_nl_cmd_get (which is the length of the returned buffer) is equal to sizeof(params), and then you copy out sizeof(params) with an offset of 4 (NL_ATTR_HEADER). So you're guaranteeing an out of bounds access. So either you need to add NL_ATTR_HDR to the check above, or you have to copy without offset.
> 
> If the offset is needed, it is going to be needed for all calls of dwpal_nl_cmd_get. So you'd better skip the header inside the dwpal_nl_cmd_get function then. But then, the copy is again redundant, you can just pass params and sizeof(params) as arguments to dwpal_nl_cmd_get.
> 
> In conclusion, I don't think this copy will be needed.
> 
> BTW I have the feeling I've already made this comment in the original #619.

changed to std::copy  and removed the safelibc commit  (no need for it now).
fixed the offset issue.

In general i prefer to leave dwpal_nl_cmd_get raw as possible and not to remove the offset there just so i could pass the params instead of the buffer.